### PR TITLE
Add generalized cache handling

### PR DESCRIPTION
**Pull request by z4kn4fein**
### Describe the purpose of your pull request

This PR enables the [generalized caching](https://configcat.com/docs/advanced/caching/#shared-cache) feature.

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

**Review by laliconfigcat — APPROVED** [review]

File: src/Cache/ConfigEntry.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $timePos = strpos($cached, "\n");
+        $etagPos = strpos($cached, "\n", $timePos + 1);
+
+        if ($timePos === false || $etagPos === false) {
+            throw new UnexpectedValueException("Number of values is fewer than expected.");
+        }
+
+        $fetchTimeString = substr($cached, 0, $timePos);
+        $fetchTime = intval($fetchTimeString);
```
Comment (adams85) [inline src/Cache/ConfigEntry.php]: Won't be this conversion problematic on 32-bit systems? The [PHP manual](https://www.php.net/manual/en/function.intval.php) says:

> 32 bit systems have a maximum signed integer range of -2147483648 to 2147483647. So for example on such a system, intval('1000000000000') will return 2147483647. The maximum signed integer value for 64 bit systems is 9223372036854775807.

A floating point number looks safer to store fetch time.

Comment (z4kn4fein) [reply src/Cache/ConfigEntry.php]: It won't be for the next 15 years 😄 

Anyway, I changed it to float!

Comment (adams85) [reply src/Cache/ConfigEntry.php]: I'm afraid that would be true if the cache entry stored the fetch time as seconds. But, we made that to store it as milliseconds, so we already have problems now...

Comment (z4kn4fein) [reply src/Cache/ConfigEntry.php]: Yep, you are right, the cache entry stored it with seconds precision as `int`, but `intval()` here could have failed reading the cache on 32-bit. It's `float` now so the problem is gone and the precision is also correct.

File: src/EvaluationDetails.php [hunk 2]
```diff
@@ -14,7 +14,7 @@ public function __construct(
         private readonly ?User $user,
         private readonly bool $isDefaultValue,
         private readonly ?string $error,
-        private readonly int $fetchTimeUnixSeconds,
+        private readonly float $fetchTimeUnixSeconds,
```
Comment (adams85) [inline src/EvaluationDetails.php]: At [ConfigCatClient.php:483](https://github.com/configcat/php-sdk/pull/36/commits/c1ec87685cf4d29a1abfb552f4b085c354e5a40c) looks like a msec value is passed but this field is second-based.

File: src/ConfigCatClient.php [hunk 3]
```diff
@@ -541,7 +541,7 @@ private function getSettingsResult(): SettingsResult
     private function getRemoteSettingsResult(): SettingsResult
     {
         $cacheEntry = $this->cache->load($this->cacheKey);
-        if (!$this->offline && $cacheEntry->getFetchTime() + $this->cacheRefreshInterval < time()) {
+        if (!$this->offline && $cacheEntry->getFetchTime() + $this->cacheRefreshInterval < Utils::getUnixMilliseconds()) {
```
Comment (adams85) [inline src/ConfigCatClient.php]: While `getFetchTime` returns msec now, `cacheRefreshInterval` is a second-based value, isn't it?

**Review by adams85 — CHANGES_REQUESTED** [review]

File: src/EvaluationDetails.php (hunk 2 above)
Comment (z4kn4fein) [reply src/EvaluationDetails.php]: 👍

File: src/ConfigCatClient.php (hunk 3 above)
Comment (z4kn4fein) [reply src/ConfigCatClient.php]: 👍

**Review by adams85 — APPROVED** [review]

File: src/EvaluationDetails.php [hunk 4]
```diff
@@ -87,11 +87,11 @@ public function getError(): ?string
     }
 
     /**
-     * @return int the last download time of the current config in unix seconds format.
+     * @return float the last download time of the current config in unix milliseconds format.
      */
-    public function getFetchTimeUnixSeconds(): int
+    public function getFetchTimeUnixMilliseconds(): float
```
Comment (adams85) [inline src/EvaluationDetails.php:92]: This is a BC to a public API, so I guess it should be mentioned in the release notes.

**Pull request merged**